metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java

```diff
@@ -0,0 +1,231 @@
+    sourceObject.put("entityType", edge.getSource().getEntityType());
+
+    final ObjectNode destinationObject = JsonNodeFactory.instance.objectNode();
+    destinationObject.put("urn", edge.getDestination().toString());
+    destinationObject.put("entityType", edge.getDestination().getEntityType());
+
+    searchDocument.set("source", sourceObject);
+
+    searchDocument.set("destination", destinationObject);
+
+    searchDocument.put("relationshipType", edge.getRelationshipType());
+
+    return searchDocument.toString();
+  }
+
+  private String toDocId(@Nonnull final Edge edge) {
+    String rawDocId =
+        edge.getSource().toString() + DOC_DELIMETER + edge.getRelationshipType() + DOC_DELIMETER + edge.getDestination().toString();
+
+    try {
+      byte[] bytesOfRawDocID = rawDocId.getBytes("UTF-8");
+      MessageDigest md = MessageDigest.getInstance("MD5");
+      byte[] thedigest = md.digest(bytesOfRawDocID);
+      return thedigest.toString();
+    } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
+      e.printStackTrace();
+      return rawDocId;
+    }
+  }
+
+  public void addEdge(@Nonnull final Edge edge) {
+    String docId = toDocId(edge);
+    String edgeDocument = toDocument(edge);
+    _graphWriteDAO.upsertDocument(edgeDocument, docId);
+  }
+
+  @Nonnull
+  public List<String> findRelatedUrns(
+      @Nullable final String sourceType,
+      @Nonnull final Filter sourceEntityFilter,
+      @Nullable final String destinationType,
+      @Nonnull final Filter destinationEntityFilter,
+      @Nonnull final List<String> relationshipTypes,
+      @Nonnull final RelationshipFilter relationshipFilter,
+      final int offset,
+      final int count) {
+
+    final RelationshipDirection relationshipDirection = relationshipFilter.getDirection();
+    String destinationNode = relationshipDirection == RelationshipDirection.OUTGOING ? "destination" : "source";
+
+    SearchResponse response = _graphReadDAO.getSearchResponse(
+        sourceType,
+        sourceEntityFilter,
+        destinationType,
+        destinationEntityFilter,
+        relationshipTypes,
+        relationshipFilter,
+        offset,
+        count
+    );
+
+    return Arrays.stream(response.getHits().getHits())
+        .map(hit -> ((HashMap<String, String>) hit.getSourceAsMap().getOrDefault(destinationNode, EMPTY_HASH)).getOrDefault("urn", null))
+        .filter(urn -> urn != null)
+        .collect(Collectors.toList());
+  }
+
+  private Filter createUrnFilter(@Nonnull final Urn urn) {
+    Filter filter = new Filter();
+    CriterionArray criterionArray = new CriterionArray();
+    Criterion criterion = new Criterion();
+    criterion.setCondition(Condition.EQUAL);
+    criterion.setField("urn");
+    criterion.setValue(urn.toString());
+    criterionArray.add(criterion);
+    filter.setCriteria(criterionArray);
+
+    return filter;
+  }
+
+  public void removeNode(@Nonnull final Urn urn) {
```

Comment:
Realized https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html is supported in open distro as well. This is likely more efficient and clean. Possibly an "or" query  which matches urn on either source or destination.

Reply:
https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html

Reply:
oh great- thanks for finding!